File: app/src/main/java/com/cristhian/gametobeat/ui/detail/GameDetailActivity.java
package com.cristhian.gametobeat.ui.detail;

import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.cristhian.gametobeat.R;
import com.cristhian.gametobeat.database.Game;
import com.cristhian.gametobeat.database.GameViewModel;
import com.cristhian.gametobeat.util.IgdbUtil;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class GameDetailActivity extends AppCompatActivity {

    public static final String PARAM_GAME_DETAIL = "param_game_detail";

    @BindView(R.id.tv_name)
    TextView mName;

    @BindView(R.id.iv_cover)
    ImageView mCover;

    @BindView(R.id.tv_summary)
    TextView mSummary;

    @BindView(R.id.tv_rating)
    TextView mRating;

    @BindView(R.id.bt_beat)
    Button mBeat;

    @BindView(R.id.bt_playing)
    Button mPlaying;

    @BindView(R.id.bt_want)
    Button mWant;

    private Game mGame;

    private GameViewModel mGameViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_detail);
        ButterKnife.bind(this);

        mGameViewModel = ViewModelProviders.of(this).get(GameViewModel.class);

        if (getIntent().getExtras() != null
                && getIntent().getExtras().containsKey(PARAM_GAME_DETAIL)) {

            mGame = getIntent().getExtras().getParcelable(PARAM_GAME_DETAIL);

            mGameViewModel.loadGameById(mGame).observe(this, new Observer<Game>() {
                @Override
                public void onChanged(@Nullable Game game) {
                    if (game != null) {
                        mGame = game;
                    }

                    setupUI();
                }
            });
        }
    }

    private void setupUI() {
        mName.setText(mGame.getName());

        mBeat.setBackgroundColor(getResources().getColor(R.color.colorAccent));
        mPlaying.setBackgroundColor(getResources().getColor(R.color.colorAccent));
        mWant.setBackgroundColor(getResources().getColor(R.color.colorAccent));

        switch (mGame.getStatus()) {
            case Game.STATUS_BEAT:
                mBeat.setBackgroundColor(getResources().getColor(R.color.colorPrimary));
                break;
            case Game.STATUS_PLAYING:
                mPlaying.setBackgroundColor(getResources().getColor(R.color.colorPrimary));
                break;
            case Game.STATUS_WANT:
                mWant.setBackgroundColor(getResources().getColor(R.color.colorPrimary));
                break;
        }

        if (mGame.getCover() != null) {
            Glide.with(this).load(IgdbUtil.getUrlImage(mGame.getCover().getImageId())).into(mCover);
        }

        if (mGame.getSummary() != null) {
            mSummary.setText(mGame.getSummary());
        }

        if (mGame.getRating() != null) {
            String ratingFormatted = String.format("%.2f", mGame.getRating());
            mRating.setText(ratingFormatted);
        }


    }

    @OnClick(R.id.bt_beat)
    public void beat() {
        changeStatus(Game.STATUS_BEAT);
    }

    @OnClick(R.id.bt_playing)
    public void playing() {
        changeStatus(Game.STATUS_PLAYING);
    }

    @OnClick(R.id.bt_want)
    public void want() {
        changeStatus(Game.STATUS_WANT);
    }

    private void changeStatus(int status) {

        if (mGame.getStatus() == status) {
            mGame.setStatus(Game.STATUS_NULL);
            mGameViewModel.delete(mGame);
        } else if (mGame.getStatus() != Game.STATUS_NULL) {
            mGame.setStatus(status);
            mGameViewModel.update(mGame);
        } else {
            mGame.setStatus(status);
            mGameViewModel.insert(mGame);
        }
    }
}

File: app/src/main/java/com/cristhian/gametobeat/ui/detail/GameDetailActivity.kt
package com.cristhian.gametobeat.ui.detail

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.cristhian.gametobeat.R
import com.cristhian.gametobeat.database.Game
import com.cristhian.gametobeat.database.GameViewModel
import com.cristhian.gametobeat.util.IgdbUtil
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_game_detail.*

class GameDetailActivity : AppCompatActivity() {

    companion object {
        const val PARAM_GAME_DETAIL = "param_game_detail"
    }

    lateinit var mGame: Game
    lateinit var mGameViewModel: GameViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game_detail)

        mGameViewModel = ViewModelProvider(this).get(GameViewModel::class.java)

        if (intent.extras != null
                && intent.extras.containsKey(PARAM_GAME_DETAIL)
        ) {
            mGame = intent.extras.getParcelable(PARAM_GAME_DETAIL)
            mGameViewModel.loadGameById(mGame)?.observe(this,
                    Observer { game ->
                        if (game != null) {
                            mGame = game
                        }
                        setupUI()
                    })
        }

        bt_beat.setOnClickListener {
            changeStatus(Game.STATUS_BEAT)
        }

        bt_playing.setOnClickListener {
            changeStatus(Game.STATUS_PLAYING)
        }

        bt_want.setOnClickListener {
            changeStatus(Game.STATUS_WANT)
        }
    }

    private fun setupUI() {
        tv_name.text = mGame.name
        bt_beat.setBackgroundColor(resources.getColor(R.color.colorAccent))
        bt_playing.setBackgroundColor(resources.getColor(R.color.colorAccent))
        bt_want.setBackgroundColor(resources.getColor(R.color.colorAccent))
        when (mGame.status) {
            Game.STATUS_BEAT -> bt_beat.setBackgroundColor(resources.getColor(R.color.colorPrimary))
            Game.STATUS_PLAYING -> bt_playing.setBackgroundColor(resources.getColor(R.color.colorPrimary))
            Game.STATUS_WANT -> bt_want.setBackgroundColor(resources.getColor(R.color.colorPrimary))
        }

        mGame.cover?.let {
            Picasso.get().load(IgdbUtil.getUrlImage(it.imageId)).into(iv_cover)
        }

        mGame.summary?.let {
            tv_summary.text = mGame.summary
        }

        mGame.rating?.let {
            val ratingFormatted = String.format("%.2f", mGame.rating)
            tv_rating.text = ratingFormatted
        }
    }

    private fun changeStatus(status: Int) {
        when {
            mGame.status == status -> {
                mGame.status = Game.STATUS_NULL
                mGameViewModel.delete(mGame)
            }
            mGame.status != Game.STATUS_NULL -> {
                mGame.status = status
                mGameViewModel.update(mGame)
            }
            else -> {
                mGame.status = status
                mGameViewModel.insert(mGame)
            }
        }
    }
}